client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java

```diff
@@ -0,0 +1,197 @@
+/**
+ * Copyright (c) 2017 Bosch Software Innovations GmbH.
+ * <p>
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ * <p>
+ * Contributors:
+ * Bosch Software Innovations GmbH - initial creation
+ */
+package org.eclipse.hono.client.impl;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.proton.*;
+import org.apache.qpid.proton.amqp.messaging.AmqpValue;
+import org.apache.qpid.proton.amqp.messaging.ApplicationProperties;
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.client.RequestResponseClient;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.RequestResponseResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A Vertx-Proton based parent class for the implementation of API clients that follow the request response pattern.
+ * The class is a generic that expects two classes:
+ * - C : denotes the concrete interface for the API
+ * - R : denotes the concrete result container class of the API
+ * Both classes have their own parent and need to be subclassed.
+ *
+ * The subclasses only need to implement some abstract helper methods (see the method descriptions) and their own
+ * API specific methods. This allows for implementation classes that focus on the API specific code.
+ */
+public abstract class AbstractRequestResponseClient<C extends RequestResponseClient,R extends RequestResponseResult>
+        extends AbstractHonoClient implements RequestResponseClient {
+
+    private static final Logger           LOG = LoggerFactory.getLogger(AbstractRequestResponseClient.class);
+    protected final AtomicLong            messageCounter  = new AtomicLong();
+    private final String                  requestResponseAddressTemplate;
+    private final String                  requestResponseReplyToAddressTemplate;;
+
+    protected final Map<String, Handler<AsyncResult<R>>> replyMap = new ConcurrentHashMap<>();
+    protected final String                replyToAddress;
+
+    /**
+     * Get the name of the endpoint that this client targets at.
+     * @return The name of the endpoint for this client.
+     */
+    protected abstract String getName();
+
+    /**
+     * Build a unique messageId for a request that serves as an identifier for a new message.
+     * @return The unique messageId;
+     */
+    protected abstract String createMessageId();
+
+    /**
+     * Construct a result object that contains the answer for a message.
```

Comment:
*Creates a result for the status and payload from a response received from the endpoint.*?